test/integration/scheduler_perf/scheduler_perf.go

```diff
@@ -988,6 +992,144 @@ func (scm stopCollectingMetricsOp) patchParams(_ *workload) (realOp, error) {
+// The check is done by listing all pods in the namespace and checking their status.
+type checkPodScheduledOp struct {
+	// Must be checkPodScheduledOpcode.
+	Opcode operationCode
+	// Namespace where the pods are located.
+	Namespace string
+	// Duration for the test.
+	Duration metav1.Duration
+	// DurationParam is the parameter for the duration.
+	// If set, the duration is parameterized through this parameter.
+	DurationParam string
+	// ScheduledParam is the parameter for the expected scheduled state.
+	// If set, the expected scheduled state is parameterized through this parameter.
+	ScheduledParam string
+	// ExpectedScheduled is the expected scheduled state.
+	// If true, the op checks if all pods are scheduled.
+	// If false, the op checks if all pods are unscheduled.
+	// This is used to test the device binding conditions.
+	ExpectedScheduled bool
+}
+
+func (op *checkPodScheduledOp) isValid(allowParameterization bool) error {
+	if op.Namespace == "" {
+		return fmt.Errorf("namespace must be set")
+	}
+	return nil
+}
+
+func (op *checkPodScheduledOp) collectsMetrics() bool {
+	return false
+}
+
+func (op *checkPodScheduledOp) patchParams(w *workload) (realOp, error) {
+	if op.ScheduledParam != "" {
+		var err error
+		op.ExpectedScheduled, err = getParam[bool](w.Params, op.ScheduledParam[1:])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if op.DurationParam != "" {
+		durationStr, err := getParam[string](w.Params, op.DurationParam[1:])
+		if err != nil {
+			return nil, err
+		}
+		if op.Duration.Duration, err = time.ParseDuration(durationStr); err != nil {
+			return nil, fmt.Errorf("parsing duration parameter %s: %w", op.DurationParam, err)
+		}
+	}
+
+	return op, op.isValid(false)
+}
+
+func (op *checkPodScheduledOp) requiredNamespaces() []string {
+	if op.Namespace != "" {
+		return []string{op.Namespace}
+	}
+	return nil
+}
+
+func (op *checkPodScheduledOp) run(tCtx ktesting.TContext) {
+	ctx, cancel := context.WithTimeout(tCtx, op.Duration.Duration)
+	defer cancel()
+
+	var (
+		lastViolation string
+		stableStart   *time.Time
+	)
+
+	validate := func(pod v1.Pod) bool {
+		isScheduled := pod.Spec.NodeName != ""
+		return (op.ExpectedScheduled && isScheduled) || (!op.ExpectedScheduled && !isScheduled)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			if ctx.Err() == context.DeadlineExceeded {
+				if op.ExpectedScheduled {
+					tCtx.Fatalf("Timeout waiting for pods to be scheduled. Context error: %v. Last violation: %s", ctx.Err(), lastViolation)
+				} else if lastViolation != "" {
+					tCtx.Fatalf("Expected pods to remain unscheduled, but some pods were scheduled. Last violation: %s", lastViolation)
+				} else {
+					// If ExpectedScheduled is false (pods should be unscheduled), timeout means success.
+					tCtx.Logf("All pods remained unscheduled for the duration")
+					return
+				}
+			} else {
+				tCtx.Fatalf("Context cancelled: %v. Last violation: %s", ctx.Err(), lastViolation)
+			}
```

Comment:
You could flatten the conditions:
```suggestion
			if ctx.Err() != context.DeadlineExceeded {
				tCtx.Fatalf("Context cancelled: %v. Last violation: %s", ctx.Err(), lastViolation)
			}
			if op.ExpectedScheduled {
				tCtx.Fatalf("Timeout waiting for pods to be scheduled. Context error: %v. Last violation: %s", ctx.Err(), lastViolation)
			}
			if lastViolation != "" {
				tCtx.Fatalf("Expected pods to remain unscheduled, but some pods were scheduled. Last violation: %s", lastViolation)
			}
			// If ExpectedScheduled is false (pods should be unscheduled), timeout means success.
			tCtx.Logf("All pods remained unscheduled for the duration")
			return
```